### ISSUE TITLE
emcc -dM -E - breaks since b57ccfef (fix for #3365)

### ISSUE THREAD

**Issue by zougloub** [emscripten-core/emscripten#3552]
Hi,

Usage of `$CC -dM -E -x c /dev/null`, `$CC -dM -E -` is standard in order to read compiler defines.
This was working until commit `b57ccfef` was introduced.

**Comment by kripken** [comment]
Do you have a testcase showing the issue? Apparently we don't have coverage for this in our test suite.

**Comment by zougloub** [comment]
Quickly added that test code!

**Comment by kripken** [comment]
(I pushed a commit to fix the error code, thanks.)

Regarding the first, the real issue is stdin, isn't it? Does "-" tell the compiler to read from stdin?

The second would seem to imply that we should assume all arguments are input files, unless told otherwise, and not look at the suffixes, which might be a big change in emcc. I wonder if it's worth it - how important is this?

**Comment by zougloub** [comment]
Yes `-` tells the compiler to read from stdin; using `/dev/stdin` would do the same thing (I didn't add a test for the latter).

Dumping compiler macros is a pretty standard way to inquire about the compiler and its target platform, and most compilers support it: http://nadeausoftware.com/articles/2011/12/c_c_tip_how_list_compiler_predefined_macros and without having to create a dummy.c file.

So IMHO this is pretty important to be able to restore that useful functionality...

**Comment by dxli** [comment]
Please fix this!

Thanks!

**Comment by mtango** [comment]
+1

**Comment by kripken** [comment]
It still seems to me like there are 2 issues here, as mentioned in my previous comment. The first (support `-` to mean stdin) seems easy to fix, the second (accept all input args as input files) seems harder or perhaps impossible without a major rewrite. Are both important to fix?

**Comment by zougloub** [comment]
Hi Alon,

Oops, I believed I had answered your questions... 

Well, you could split the issue in two if you want to.
The support of `-` to mean stdin is one, so `$CC -dM -E -` works is an immediate problem that *I* have, so I'd be glad if it's fixed.
But I'm also convince that emcc will have trouble in the long term if it doesn't do the second one (as gcc and clang do... they have assumption on file type regarding extensions, but it can be overridden, and compile different files into one lib/executable `gcc -o ab a.c b.c && clang -o ab.so a.c b.c -shared`).

**Comment by kripken** [comment]
Ok, I looked into this a little more.

 * Supporting `-` means also noticing `-x` in order to know what the input is, if we take the easy route of reading stdin to a file. Unless we defer it all to clang, which requires some surgery.
 * Supporting `/dev/null` etc. as inputs means we do not trust file suffixes. If we just assume they are source files, we do ok on `/dev/null` but badly on `libsomething.so.1.0.6`, for example.

So neither of those is trivial to do, I'm afraid. I do agree it would be good to fix this, though, if someone wants to try.

Reverting b57ccfef is not an option, as we never really worked on this - we just happened to have a hack that worked properly on that input. Perhaps we can find another hack to get it to work again somehow, like seeing early on (as we did before that commit) that it is ok to just do the previous behavior (call clang and quit)? Not obvious to me what the decision would be made based on, though.
